### DISCUSSION TITLE
JS fractions for big numbers with no precision lose

### DISCUSSION THREAD

**Discussion by valerii15298** [josdejong/mathjs#2957] [Q&A] [answered]
Big numbers with fraction support:

I want to perform math operation with no precision loss.
How can I do it using JavaScript?
First I tried [decimal.js](https://github.com/MikeMcl/decimal.js/), but it does not preserve precision as you see in next example:
```ts
import {Decimal} from 'decimal.js';
const a1 = new Decimal(1);
const a2 = new Decimal(3);
console.log(a1.div(a2).toFraction()); // should be `1/3` but I get `33333333333333333333/100000000000000000000`
```

Next I tried [fraction.js](https://github.com/infusion/Fraction.js) but it does not work with large numbers, for example:
```ts
import Fraction from 'fraction.js';
const n1 = new Fraction("99999999999999999999999999999999999999999999999");
const n2 = new Fraction("99999999999999999999999999999999999999999999990");
console.log(+(n1.sub(n2))) // should be 9, but I get 0
```

Are there any solutions to work with relative large numbers(lets say same as `decimal.js` supports) but with high precision(same as `fraction.js` supports).

I checked out that `mathjs` uses `fraction.js` under the hood, so no any advantages for large numbers:
```ts
import * as math from "mathjs";
const a1 = math.fraction(math.number("99999999999999999999999999999999999999999999999"))
const a2 = math.fraction(math.number("99999999999999999999999999999999999999999999990"))
console.log(+(math.subtract(a1, a2))) // should be 9 but I get 0
```

**Comment by josdejong** [comment]
> I want to perform math operation with no precision loss.

Please ask yourself _why_ you think you need that. 

There is no silver bullet. There are roughly the following options:

1. Using Floating point `number` has a precision of about 16 digits
2. Using `BigNumber` (powered by decimal.js) has a configurable precision, but still is floating point and has a finite precision so it cannot store a division `1/3` without loss nor `pi` with endless precision.
3. Using `Fraction` can store a division like 1/3 without loss of precision, but you can't store `pi` that way, nor do operations like `sqrt` and `sin`
4. Using algebra and symbolic computation. mathjs has the function `simplify` that can be powerful. It has it's limitations though.

Docs: https://mathjs.org/docs/datatypes/numbers.html#roundoff-errors

On a side note: your last example creates a regular `number` from a string like `"99999999999999999999999999999999999999999999999"`, but a `number` can  only hold about 16 digits, so that is not helpful. In mathjs you can use `math.bignumber` to create a `BigNumber`, it is configured with a precision of 64 digits by default (and configurable). You can't combine `BigNumber` with `math.fraction` though.

**Reply by valerii15298** [reply]
Yeah, that is basically my problem, to combine BigNumber with Fraction...

Btw, very thanks for a such detailed explanatio!!

The point is I am working with financial app, where I do not wanna lose any precision. I do not need any fancy stuff like sqrt, etc...

All I need is add, subtract, multiply and unfortunately divide which creates a problem for me to use Fraction... But with Fraction I cannot use BigNumber... so I should either to build my own small implementation of Fraction with BigInt or look at something like https://github.com/infusion/Fraction.js/blob/master/bigfraction.js the last one unfortunately doesn't not have any docs in their repo so I'm not sure about it.
If you have any recommendations it will be very helpful. Thanks again!!!

**Reply by josdejong** [reply]
That makes sense. I do not know any details of your case, but I think I would use a `BigNumber` configured with a large enough precision.

First I would set up a set of realistic calculations and the expected result I.e. `1/3` in itself is not a real use case. A real use case is probably something like multiply an amount of money with 2 digits with some interest rate with 5 digits, and the end result probably only needs 2 or 3 digits precision or so. You can do calculations with high floating point precision, and then round the end result to the required number of digits to get rid of any round-off errors.

**Reply by josdejong** [reply, marked as answer]
A `BigFraction` library would be very nice too, given that all operations that you need are supported by the library. (Isn't the regular fraction.js library good enough though?... test it against your realistic calculations ;) )

**Reply by valerii15298** [reply]
It just that we are afraid 16 digits of js number type can be not enough... Thank you!!!!!!
